File: Depth_First_Search/2368.Reachable_Nodes_With_Restrictions/Python/2368.Reachable_Nodes_With_Restrictions.py
class Solution(object):
    def __init__(self):
        """
        :type: None, void
        :rtype: None, void
        """
        ############
        #Initialize
        ##### Result maximun summary counter #####
        self.res_max_cnter = 0


    def reachableNodes(self, n, edges, restricted):
        """
        :type n: int
        :type edges: List[List[int]]
        :type restricted: List[int]
        :rtype: int
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Length of edges, restricted array #####
        len_edges, len_restricted = len(edges), len(restricted)

        ##### Visit array #####
        visit_arry = [0 for _ in range(n)]

        ##### Record dictionary/hashmap #####
        record_dict = {}


        ####################################
        #Recursion-based DFS loop traversal
        for restricted_idx in range(len_restricted):
            visit_arry[restricted[restricted_idx]] = 1 #Keep updating/overwriting

        for edges_idx in range(len_edges):
            first_node, second_node = (edges[edges_idx])[0], (edges[edges_idx])[1] #First, Second-node

            ##### Check if the current node value existed or not #####
            if (first_node not in record_dict):
                record_dict[first_node] = [second_node] #Keep updating/recording
            else:
                (record_dict[first_node]).append(second_node) #Keep updating/recording

            if (second_node not in record_dict):
                record_dict[second_node] = [first_node] #Keep updating/recording
            else:
                (record_dict[second_node]).append(first_node) #Keep updating/recording

        self.dfsTraversal(record_dict, visit_arry, 0) #Recursion function call

        return self.res_max_cnter


    def dfsTraversal(self, record_dict, visit_arry, record_val):
        """
        :type record_dict: dict
        :type visit_arry: List[int]
        :type record_val: int
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (visit_arry[record_val]): #Issue/Boundary-case handling
            return

        self.res_max_cnter += 1 #Update/Accumulate

        visit_arry[record_val] = 1 #Update/Overwrite

        
        ####################
        #Whole process/flow
        ##### Check if the current node-value existed or not #####
        if (record_val in record_dict):
            for next_val in record_dict[record_val]:
                self.dfsTraversal(record_dict, visit_arry, next_val) #Recursion function call

File: Depth_First_Search/2368.Reachable_Nodes_With_Restrictions/C++/2368.Reachable_Nodes_With_Restrictions.cpp
class Solution {
public:
    int reachableNodes(int n, vector<vector<int>> & edges, vector<int> & restricted) 
    {
        //=============================================//
        // Input type:                                 //
        //   - int (n-th nodes)                        //
        //   - vector<vector<int>> ref. (edges vector) //
        //   - vector<int> ref. (restricted vector)    //
        // Return type:                                // 
        //   - int (result maximun summary value)      //
        //=============================================//

        /*Initialize*/
        ///// Lenght of vectors (i.e. edges, restricted) /////
        int len_edges = edges.size(), len_restricted = restricted.size();

        ///// Visit vector /////
        vector<int> visit_vect(n, 0);

        ///// Root-child dictionary, hashmap /////
        unordered_map<int, vector<int>> root_child_map;

        ///// Result maximun summary value /////
        int res_max_sum = 0;


        /*Recursion-based DFS loop traversal*/
        for (int restricted_idx = 0; (restricted_idx < len_restricted); (restricted_idx++)) //Whole
        {
            visit_vect[restricted[restricted_idx]] = 1; //Keep updating, overwriting

        } //Whole

        for (int edges_idx = 0; (edges_idx < len_edges); (edges_idx++)) //Whole
        {
            (root_child_map[(edges[edges_idx])[0]]).emplace_back((edges[edges_idx])[1]); //Keep updating, recording
            (root_child_map[(edges[edges_idx])[1]]).emplace_back((edges[edges_idx])[0]); //Keep updating, recording

        } //Whole

        dfsTraversal(root_child_map, visit_vect, 0, res_max_sum); //Recursion function call

        return res_max_sum;
    }


    void dfsTraversal(unordered_map<int, vector<int>> & root_child_map, vector<int> & visit_vect, int record_val, int & res_max_sum)
    {
        //===========================================================================//
        // Input type:                                                               //
        //   - unordered_map<int, vector<int>> ref. (root-child dictionary, hashmap) //
        //   - vector<vector<int>> ref. (visit vector)                               //
        //   - int (record node-value)                                               //
        //   - int ref. (result maximun summary value)                               //
        // Return type:                                                              // 
        //   - void (no return)                                                      //
        //===========================================================================//
        if (visit_vect[record_val]) { return; } //Issue, Boundary-case handling

        (res_max_sum++); visit_vect[record_val] = 1; //Update, Overwrite

        for (const int & next_val: root_child_map[record_val]) //Whole
        {
            dfsTraversal(root_child_map, visit_vect, next_val, res_max_sum); //Recursion function call

        } //Whole
    }
};
